core/src/main/scala/org/apache/spark/network/SyncBlockTransferClient.scala

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.network
+
+import java.nio.ByteBuffer
+
+import scala.concurrent.Promise
+import scala.concurrent.duration.Duration
+
+import org.apache.spark.network.buffer.{FileSegmentManagedBuffer, ManagedBuffer, NioManagedBuffer}
+import org.apache.spark.network.shuffle.{BlockFetchingListener, BlockTransferClient, DownloadFileManager}
+import org.apache.spark.storage.EncryptedManagedBuffer
+import org.apache.spark.util.ThreadUtils
+
+private[spark] class SyncBlockTransferClient(val blockTransferClient: BlockTransferClient) {
```

Comment:
So I'm trying to understand why you need this class and the new interface. I read your description but it didn't help.

All the places where I see this class instantiated already have a reference to `BlockTransferService`, which is where `fetchBlockSync` currently lives. Could you explain why you need the new `BlockTransferClient` and this new class instead of just using `BlockTransferService` as is?

Reply:
When the external shuffle service is enabled we are creating an external shuffle client (in contrast to `BlockTransferService`) for fetching the blocks:

https://github.com/apache/spark/blob/df3a80da4270c7b5eddb83383d8149ed64b25a66/core/src/main/scala/org/apache/spark/storage/BlockManager.scala#L178-L186  

And `ExternalShuffleClient` is inherited directly from `ShuffleClient`. Likewise `BlockTransferService` so they are "just" siblings and the common functionality must be at least at their parent (`ShuffleClient`) level. 

During the development once I had a version where `ShuffleClient` was used within `SyncBlockTransferClient` instead of `BlockTransferClient` (this way I saved that interface) but for understanding point of view I took that very misleading so I decided to keep `BlockTransferClient`.

Reply:
There are a few reasons why I don't like this class very much. First, it's called `SyncBlockTransferClient` but it's not a `BlockTransferClient`. But the main one is that it's basically a helper method. You could have `fetchBlockSync` implemented in `ShuffleClient`, for example, and it would be less code overall without the weirdness.

As for the interface, I see the point of `BlockTransferClient` being more generic than `ShuffleClient`, but at the same time, aren't all current implementations extending `ShuffleClient` anyway? It's more of a naming thing than a necessity as far as I can see. You also end up with a weird inheritance chain (`BlockTransferService` is a `BlockTransferClient`). Part of the naming is because the existing names are not great, but I'm not seeing an advantage in adding more to the name salad here.

Reply:
> You could have fetchBlockSync implemented in ShuffleClient, for example, and it would be less code overall without the weirdness.

In the description I mentioned some problems with this:
> ShuffleClient is in the spark-network-shuffle artifact where neither EncryptedManagedBuffer and SparkException (used in the awaitResult() which should have been reimplemented here with a Java promise) are not available.

Reply:
Well, it could still just be a helper method somewhere, not a full class that doesn't do anything else.

Reply:
I made a mistake here. During development in one of the test I was using `ShuffleClient` for accessing the blocks synchronously. That test finally evolved the more higher level test 
 which is in the `ExternalShuffleServiceSuite`. And now I see it is really enough for the `BlockTransferService` to be able to transform the asynchronous fetch to a synch fetch. So I revert the `SyncBlockTransferClient` changes.